crates/loro-internal/src/kv_store.rs

```diff
@@ -110,110 +117,1619 @@ mod default_binary_format {
+                self.data.len() +
+                // checksum
+                SIZE_OF_U32
+            }
+        }
+
+        pub fn is_empty(&self)->bool{
+            !self.is_large && self.offsets.is_empty()
+        }
+
+        /// Add a key-value pair to the block.
+        /// Returns true if the key-value pair is added successfully, false the block is full.
+        /// 
+        /// ┌───────────────────────────────────────────────────────────────┐
+        /// │  Key Value Chunk                                              │
+        /// │┌ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ┬ ─ ─ ─ ┐│
+        /// │ common prefix len key suffix len│key suffix│value len  value  │
+        /// ││       u8        │     u16      │  bytes   │   u16   │ bytes ││
+        /// │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ │
+        /// └───────────────────────────────────────────────────────────────┘
+        /// 
+        pub fn add(&mut self, key: &[u8], value: &[u8]) -> bool {
+            debug_assert!(!key.is_empty(), "key cannot be empty");
+            if  self.first_key.is_empty() && value.len() > self.block_size {
+                let key_len = key.len() as u16;
+                self.data.put_u16(key_len);
+                self.data.put(key);
+                self.data.put(value);
+                self.is_large = true;
+                self.first_key = key.to_vec();
+                return true;
+            }
+
+            // whether the block is full
+            if self.estimated_size() + key.len() + value.len() + SIZE_OF_U8 + SIZE_OF_U16 * 2 > self.block_size && !self.first_key.is_empty() {
+                return false;
+            }
+
+            self.offsets.push(self.data.len() as u16);
+            let (common, suffix) = get_common_prefix_len_and_strip(key, &self.first_key);
+            let key_len = suffix.len() as u16;
+            let value_len = value.len() as u16;
+            self.data.put_u8(common);
+            self.data.put_u16(key_len);
+            self.data.put(suffix);
+            self.data.put_u16(value_len);
+            self.data.put(value);
+            if self.first_key.is_empty() {
+                self.first_key = key.to_vec();
+            }
+            true
+        }
+
+        pub fn build(self)->Block{
+            if self.is_large{
+                return Block::Large(LargeValueBlock{
+                    data: Bytes::from(self.data),
+                    key_length: self.first_key.len(),
+                });
+            }
+            debug_assert!(!self.offsets.is_empty(), "block is empty");
+            Block::Normal(NormalBlock{
+                data: Bytes::from(self.data),
+                offsets: self.offsets,
+            })
+        }
+    }
+
+    /// ┌──────────────────────────────────────────────────────────────────────────────────────┐
+    /// │ Block Meta                                                                           │
```

Comment:
We don't need to store first key and last key in the block meta. It can be read from the block directly.

And I think maybe we can also reuse the `is_large` byte to store extra info like how the block is compressed

Reply:
Oh, there is also a large block.  Yes, we should parse the key from the block directly.